src/btree/bt_page.c

```diff
@@ -120,249 +161,396 @@ __page_build_ref(WT_SESSION_IMPL *session, WT_REF *parent_ref, WT_CELL_UNPACK_AD
-              session, ref, base_key, base_val, NULL, true, &refs[final_entries++], incr));
-        } else if (cmp >= 0) {
-            if (!__wt_delta_cell_type_visible_all(delta[j]))
-                WT_ERR(__page_build_ref(
-                  session, ref, NULL, NULL, delta[j], false, &refs[final_entries++], incr));
-            if (cmp == 0)
-                i += 2; /* Skip the current key and value. */
-            j++;
+    bool row_leaf_page = false, internal_page = false;
+
+    if (build_disk) {
+        WT_ASSERT(session, disk_image != NULL);
+        /*
+         * hdr = (WT_PAGE_HEADER *)disk_image->data; p = (uint8_t *)hdr + sizeof(WT_PAGE_HEADER);
+         */
+
+        switch (page->type) {
+        case WT_PAGE_ROW_LEAF:
+            row_leaf_page = true;
+            break;
+        case WT_PAGE_ROW_INT:
+            internal_page = true;
+            break;
+        default:
+            return (__wt_illegal_value(session, page->type));
         }
-    }
-    /* Copy the remaining entries from the base array or the delta array. */
-    for (; i < base_entries;) {
+    } else {
+        /*
+         * Encode the first key always from the base image. The btrees using customized collator
+         * cannot handle the truncated first key.
+         */
         base_key = &base[i++];
         base_val = &base[i++];
         WT_ERR(__page_build_ref(
           session, ref, base_key, base_val, NULL, true, &refs[final_entries++], incr));
     }
-    for (; j < delta_entries; j++)
-        if (!__wt_delta_cell_type_visible_all(delta[j]))
-            WT_ERR(__page_build_ref(
-              session, ref, NULL, NULL, delta[j], false, &refs[final_entries++], incr));
 
-    WT_ASSERT(session, i == base_entries && j == delta_entries);
-    WT_ASSERT(session, final_entries != 0);
-    WT_ASSERT(session, final_entries < estimated_entries && refs[final_entries] == NULL);
+    /* Common merge loop */
+    for (;;) {
+        WT_CELL_UNPACK_DELTA_INT *min_delta = NULL;
+        WT_ITEM base_key_buf, delta_key_buf;
+
+        /* Use the macro instead of the function */
+        WT_ERR(__page_find_min_delta(
+          session, unpacked_deltas, delta_size_each, delta_idx, delta_size, &min_delta, &min_d));
+
+        WT_ERR(ret);
+
+        if (i >= base_entries && min_delta == NULL)
+            break;
+
+        /* Compare base and delta keys */
+        if (i >= base_entries)
+            cmp = 1;
+        else if (min_delta == NULL)
+            cmp = -1;
+        else {
+            base_key_buf.data = base[i].data;
+            base_key_buf.size = base[i].size;
+            delta_key_buf.data = min_delta->key.data;
+            delta_key_buf.size = min_delta->key.size;
+            WT_ERR(
+              __wt_compare(session, S2BT(session)->collator, &base_key_buf, &delta_key_buf, &cmp));
+        }
+
+        /* Old implementation: build WT_REFs */
+        if (!build_disk) {
+
+            if (cmp < 0) {
```

Comment:
If here we only have base or delta entries not both, how the merge work?

Reply:
If we are running out of base page key values, we should encode them after the loop. The same for deltas.